### DISCUSSION TITLE
Spring time calculation accuracy

### DISCUSSION THREAD

**Discussion by ebraminio** [cosinekitty/astronomy#189] [Q&A] [answered]
I'm a little unsure about this that's why I'm filing it as Q&A,

Spring exact moment is very important for Persian calendar and that very exact moment will make a leap year or not. That's why some of Persian calendar implementations are quite complicated https://github.com/persian-calendar/calendar/blob/main/src/main/java/io/github/persiancalendar/calendar/persian/AlgorithmicConverter.java and the same implementation is in use in .NET to provide Persian calendar but there exists simpler algorithms able to calculate this for shorter period of time.

Culturally we got Nowruz also which happen on the Spring equinox day https://en.wikipedia.org/wiki/Nowruz marking the new year of Persian calendar.

The government here each year publishes an official document containing exact Spring time along second and here is the reference numbers it gives,

```
// Year of spring equinox - month - day - hours - minutes - second
// It should be subtracted by 3:30 to become GMT time
[
  [2022, 3, 20, 19,  3, 26],
  [2021, 3, 20, 13,  7, 28]
  [2020, 3, 20,  7, 19, 37],
  [2019, 3, 21,  1, 28, 27],
  [2018, 3, 20, 19, 45, 28],
  [2017, 3, 20, 13, 58, 40],
  [2016, 3, 20,  8,  0, 12],
  [2015, 3, 21,  2, 15, 11],
  [2014, 3, 20, 20, 27,  7],
  [2013, 3, 20, 14, 31, 56],
  [2012, 3, 20,  8, 44, 27],
  [2011, 3, 21,  2, 50, 45],
  [2010, 3, 20, 21,  2, 13],
  [2009, 3, 20, 15, 13, 39],
  [2008, 3, 20,  9, 18, 19],
  [2007, 3, 21,  3, 37, 26],
  [2006, 3, 20, 21, 55, 35],
  [2005, 3, 20, 16,  3, 24],
  [2004, 3, 20, 10, 18, 37],
  [2003, 3, 21,  4, 29, 45],
  [2002, 3, 20, 22, 46,  2]
]
```

And the numbers, along the seconds are used everywhere, from TV to printed calendars to social medias and so.

As I didn't want to hard code those number so far I could calculate equinox using this code https://github.com/persian-calendar/equinox and it approaches those official number but still has an difference less than one minute. The `Astronomy.season` calculation also gives even better numbers as can be seen [here](https://github.com/persian-calendar/persian-calendar/compare/astronomy#diff-c1979f078917d055926035007efe1becc46171dd3dfcb2eb6d93e806b798d0c6L14) yet it still differs with officially declared ones.

So my question is it possible to get those exact numbers somehow or how they've calculated anyway at the first place or maybe they are inaccurate anyway? Thanks!

**Comment by cosinekitty** [comment, marked as answer]
Yes, I think if you wanted the exact numbers, you could definitely get a lot closer. Chances are they are using something like the DE405 ephemeris, which uses large data files that contain exact numeric integration of the solar system. Another possibility is, they might be using the complete VSOP87 model for the Earth, which would be much slower to calculate than either DE405 or Astronomy Engine, but would be smaller code+data than the ephemeris file. Astronomy Engine calculates the Earth's orbital position using a truncated version of VSOP87 to sacrifice a small amount of accuracy for a huge gain in speed, with much smaller code. It is possible to adjust the Astronomy Engine code generator so that it no longer truncates the VSOP87 series. I can help you try that if you want.

**Reply by ebraminio** [reply]
Using your great input I've done this:

```python
import astronomy
from skyfield import api, almanac
import datetime

ts = api.load.timescale()
s = almanac.seasons(api.load('de440.bsp'))
year = 2022
for year in range(2002, 2023):
    t0 = ts.utc(year, 3, 19)
    t1 = ts.utc(year, 3, 22)
    de440_result = almanac.find_discrete(t0, t1, s)[0].utc_datetime()[0].replace(tzinfo=None)
    astronomy_result = astronomy.Seasons(year).mar_equinox.Utc()
    print(de440_result.isoformat(), (de440_result - astronomy_result).microseconds / 10**6)
```

```
2002-03-20T19:16:08.343961 0.262566
2003-03-21T00:59:45.674457 0.283194
2004-03-20T06:48:38.059852 0.687603
2005-03-20T12:33:25.066253 0.792372
2006-03-20T18:25:33.734505 0.90947
2007-03-21T00:07:24.913092 0.641162
2008-03-20T05:48:17.445126 0.611729
2009-03-20T11:43:37.473193 0.820144
2010-03-20T17:32:12.007026 0.071793
2011-03-20T23:20:43.426693 0.940697
2012-03-20T05:14:25.662458 0.212835
2013-03-20T11:01:54.745383 0.362777
2014-03-20T16:57:05.595070 0.716266
2015-03-20T22:45:09.337961 0.094782
2016-03-20T04:30:11.161174 0.225549
2017-03-20T10:28:37.877466 0.315148
2018-03-20T16:15:27.307987 0.956386
2019-03-20T21:58:25.820789 0.454742
2020-03-20T03:49:37.119332 0.479148
2021-03-20T09:37:28.521876 0.754001
2022-03-20T15:33:24.902437 0.110811
```

which comparing the result against de405, de421 and de440 shows the difference is less than one second, which, I can't express it enough how fantastic is this and nothing can be done realistically from astronomy engine side, and we will left with one question on how that official source is doing themselves which is out of scope for the library, still.

Great help. Thank you so much! 🙏

**Reply by cosinekitty** [reply]
I'm glad you found a good solution for what you wanted. 

Astronomy Engine does not have this extreme level of accuracy because I was trying to solve a different problem. Originally, I was trying to find a way to do astronomy calculations inside the JavaScript in a web page. The only solutions were either not accurate enough, or they were too large.

For example, how large is the file `de440.bsp`?  When I looked at all these ephemeris files, they were often tens or hundreds of megabytes, which is not practical for doing client-side calculations for the browser.

The other available solution was Keplerian elements, which provide much smaller and faster code. I wanted angular accuracy on the order of 1 arcminute, which is much more accurate than Keplerian elements can handle. These elliptical approximations for planet orbits get worse over time. So I found that no matter what I did, I could not get a good compromise between small code and good enough accuracy. This is what motivated me to start working on Astronomy Engine in 2019.

I don't believe there is a single approach that will make everyone happy. My goal in Astronomy Engine is to provide a missing middle level between very small but too inaccurate, and very large and more accurate than most people need. It is certainly possible to create another very large, slow, and extremely accurate astronomy library, but I don't think it would be adding as much value to what already exists.

**Reply by ebraminio** [reply]
> For example, how large is the file de440.bsp? 

Oh I definitely don't need sub second accuracy here as our official times are off by seconds from what multiple sources are generating so I lean toward they are wrong instead. They all are here https://ssd.jpl.nasa.gov/ftp/eph/planets/bsp/ and I tested with de440 which is 114.3M and there is de441.bsp which is 3.08G.

> I was trying to find a way to do astronomy calculations inside the JavaScript in a web page.
> It is certainly possible to create another very large, slow, and extremely accurate astronomy library, but I don't think it would be adding as much value to what already exists.

This is a very good thing for mobile apps also I can say.

**Reply by cosinekitty** [reply]
Yes, small is good for web pages, mobile apps, and also for microcontroller devices (Arduino and Raspberry Pi). These are the kind of environments I had in mind for Astronomy Engine to support.
